tests/utils/command_test_utils.go

```diff
@@ -0,0 +1,253 @@
+	}
+
+	if GetRandomDecision(2, 1) {
+		compositeCommand.Group = addGroup()
+	}
+
+	if GetBinaryDecision() {
+		compositeCommand.Label = GetRandomString(12, false)
+		LogInfoMessage(fmt.Sprintf("....... label: %s", compositeCommand.Label))
+	}
+
+	if GetBinaryDecision() {
+		compositeCommand.Parallel = true
+		LogInfoMessage(fmt.Sprintf("....... Parallel: %t", compositeCommand.Parallel))
+	}
+}
+
+// VerifyCommands verifies commands returned by the parser are the same as those saved in the devfile schema
+func (devfile TestDevfile) VerifyCommands(parserCommands []schema.Command) error {
+
+	LogInfoMessage("Enter VerifyCommands")
+	var errorString []string
+
+	// Compare entire array of commands
+	if !cmp.Equal(parserCommands, devfile.SchemaDevFile.Commands) {
+		errorString = append(errorString, LogErrorMessage(fmt.Sprintf("Command array compare failed.")))
+		// Array compare failed. Narrow down by comparing indivdual commands
+		for _, command := range parserCommands {
+			if testCommand, found := getSchemaCommand(devfile.SchemaDevFile.Commands, command.Id); found {
+				if !cmp.Equal(command, *testCommand) {
+					parserFilename := AddSuffixToFileName(devfile.FileName, "_"+command.Id+"_Parser")
+					testFilename := AddSuffixToFileName(devfile.FileName, "_"+command.Id+"_Test")
+					LogInfoMessage(fmt.Sprintf(".......marshall and write devfile %s", devfile.FileName))
+					c, err := yaml.Marshal(command)
+					if err != nil {
+						errorString = append(errorString, LogErrorMessage(fmt.Sprintf(".......marshall devfile %s", parserFilename)))
+					} else {
+						err = ioutil.WriteFile(parserFilename, c, 0644)
+						if err != nil {
+							errorString = append(errorString, LogErrorMessage(fmt.Sprintf(".......write devfile %s", parserFilename)))
+						}
+					}
+					LogInfoMessage(fmt.Sprintf(".......marshall and write devfile %s", testFilename))
+					c, err = yaml.Marshal(testCommand)
+					if err != nil {
+						errorString = append(errorString, LogErrorMessage(fmt.Sprintf(".......marshall devfile %s", testFilename)))
+					} else {
+						err = ioutil.WriteFile(testFilename, c, 0644)
+						if err != nil {
+							errorString = append(errorString, LogErrorMessage(fmt.Sprintf(".......write devfile %s", testFilename)))
+						}
+					}
+					errorString = append(errorString, LogInfoMessage(fmt.Sprintf("Command %s did not match, see files : %s and %s", command.Id, parserFilename, testFilename)))
+				} else {
+					LogInfoMessage(fmt.Sprintf(" --> Command  matched : %s", command.Id))
+				}
+			} else {
+				errorString = append(errorString, LogErrorMessage(fmt.Sprintf("Command from parser not known to test - id : %s ", command.Id)))
+			}
+
+		}
+		for _, command := range devfile.SchemaDevFile.Commands {
+			if _, found := getSchemaCommand(parserCommands, command.Id); !found {
+				errorString = append(errorString, LogErrorMessage(fmt.Sprintf("Command from test not returned by parser : %s ", command.Id)))
+			}
+		}
```

Comment:
Any reason we are looping again over here? We just looped thru test devfile and parsed commands in the above for loop.

Reply:
The previous loop goes through all the commands returned by the parser and makes sure they are the same as those expected by test. This loop does a quick check that the test does not have commands that the parser did not return.